src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+
+    using Microsoft.DocAsCode.Plugins;
+
+    public class UriTemplate<T>
+    {
+        private static readonly Regex _marcoRegex = new Regex(@"{%\s*([\S]+?)\s*%}");
+        private static readonly Regex _pipelineRegex = new Regex(@"\|>\s*([^|\s]+)\s*(.*?)\s*(?:$|(?=\|>))");
+
+        private readonly string _template;
+        private readonly Func<string, T> _func;
+        private readonly IUriTemplatePipeline<T>[] _pipeline;
+        private readonly string[][] _parameters;
+
+        /// <summary>
+        /// step 1 (macro):
+        /// * env var:      {%abc%}                         -> expand env var abc.
+        /// Step 2 (pipeline):
+        /// * pipeline:     |> function parameters          -> "", and append a pipeline.
```

Comment:
>function parameters [](start = 31, length = 19)

seems function name is not used. What about move parameters into UriTemplatePipeline's ctor? Then step2 isn't needed in template string.

Reply:
Function name is the contract name of IUriTemplatePipeline\<T\>.
If move to parameters into UriTemplatePipeline's ctor, i need one more interface: `IUriTemplatePipelineFactory<T>`

Reply:
@superyyrrzz , see the create method added in unit test.